nselib/punycode.lua

```diff
@@ -0,0 +1,664 @@
+-- string of Unicode symbols.
+--
+-- @param input The Punycode string of ASCII-only symbols.
+-- @param encoder Defines the type of encoding format to be used.
+-- @return The resulting string of Unicode symbols.
+-- The following function is adapted from punycode.js by Mathias Bynens
+-- under the MIT License.
+function decode_input(input, encoder)
+
+  local output = {}
+  local inputLength = #input
+  local i = 0
+  local n = initialN
+  local bias = initialBias
+
+  local basic
+  if find(reverse(input), delimiter) then
+    basic = #input - find(reverse(input), delimiter)
+  else
+    basic = -1
+  end
+
+  if basic < 0 then
+    basic = 0
+  end
+
+  for j = 1, basic do
+    local c = sub(input, j, j)
+    local value = byte(c)
+
+    if value >= 0x80 then
+      --error("Not basic")
+      return nil, "Not basic exception occurred."
+    end
+    table.insert(output, value)
+  end
+
+  local index
+  if basic > 0 then
+    index = basic + 1
+  else
+    index = 0
+  end
+
+  while index < inputLength do
+    local oldi = i
+    local w = 1
+    local k = base
+
+    repeat
+
+      if index >= inputLength then
+        --error("Invalid input")
+        return nil, "Invalid input exception occurred."
+      end
+
+      local c = sub(input, index+1, index+1)
+      local value = byte(c)
+      local digit = basicToDigit(value)
+
+      index = index + 1
+
+      if (digit >= base or digit > floor((maxInt - i) / w)) then
+        --error('overflow');
+        return nil, "Overflow exception occurred."
+      end
+      i = i + digit * w;
+
+      local t
+      if k <= bias then
+        t = tMin
+      else
+        if k >= bias + tMax then
+          t = tMax
+        else
+          t = k - bias
+        end
+      end
+
+      if digit < t then
+        break
+      end
+
+      local baseMinusT = base - t;
+      if (w > floor(maxInt / baseMinusT)) then
+        --error('overflow');
+        return nil, "Overflow exception occurred."
+      end
+
+      w = w * baseMinusT;
+      k = k + base
+
+    until false
+
+    local out = #output + 1;
+
+    bias = adapt(i - oldi, out, oldi == 0)
+
+    -- `i` was supposed to wrap around from `out` to `0`,
+    -- incrementing `n` each time, so we'll fix that now:
+    if (floor(i / out) > maxInt - n) then
+      --error('overflow');
+      return nil, "Overflow exception occurred."
+    end
+
+    n = n + floor(i / out);
+    i = i % out;
+    for temp = #output, i, -1 do
+      output[temp+1] = output[temp]
+    end
+    output[i+1] = n
+    i = i + 1
+  end
+
+  return unicode.encode(output, encoder)
+
+end
+
+-- The following function looks for non-ASCII characters in a string.
+--
+-- @param s String of input to be encoded.
+-- @param decoder A decoder function to convert the domain into a
+-- table of Unicode code points.
+-- @return Returns encoded string.
+function encode_label(s, decoder)
+
+  local flag = false
+  local decoded_tbl = unicode.decode(s, decoder)
```

Comment:
Remember, for simplicity we can require callers of punycode functions to already do the decoding and pass in a table of code points. After all, this is what `idna.lua` has after doing the mapping/validation, so it doesn't make sense to keep encoding and decoding between calls.

Reply:
I thought decoding here will help the NSE developers who import this function exclusively.